src/command/expire/expire.c

```diff
@@ -863,6 +863,100 @@ removeExpiredBackup(InfoBackup *infoBackup, const String *adhocBackupLabel, unsi
     FUNCTION_LOG_RETURN_VOID();
 }
 
+/***********************************************************************************************************************************
+Remove expired backup history manifests from repo
+***********************************************************************************************************************************/
+static void
+removeExpiredHistory(InfoBackup *infoBackup, unsigned int repoIdx)
+{
+    FUNCTION_LOG_BEGIN(logLevelDebug);
+        FUNCTION_LOG_PARAM(INFO_BACKUP, infoBackup);
+        FUNCTION_LOG_PARAM(UINT, repoIdx);
+    FUNCTION_LOG_END();
+
+    ASSERT(infoBackup != NULL);
+
+    MEM_CONTEXT_TEMP_BEGIN()
+    {
+        if (cfgOptionIdxTest(cfgOptRepoRetentionHistory, repoIdx))
+        {
+            // Get current backups in backup.info - these will not be expired
+            StringList *currentBackupList = strLstSort(infoBackupDataLabelList(infoBackup, NULL), sortOrderDesc);
+
+            // If the full backup history manifests are expired, we should expire diff and incr too. So, format the oldest full
+            // backup label to retain.
+            const time_t minTimestamp = time(NULL) - (time_t)(cfgOptionIdxUInt(cfgOptRepoRetentionHistory, repoIdx) * SEC_PER_DAY);
+            String *minBackupLabel = backupLabelFormat(backupTypeFull, NULL, minTimestamp);
+
+            // Get all history years
+            const StringList *historyYearList = strLstSort(
+                storageListP(
+                    storageRepo(), STRDEF(STORAGE_REPO_BACKUP "/" BACKUP_PATH_HISTORY), .expression = STRDEF("^2[0-9]{3}$")),
+                sortOrderDesc);
```

Comment:
It would be more efficient if the year list was sorted Ascending. For example, if a user would like to keep the last 2 years of history then we can stop looking when we get to the current-year - 2 years. 

Also, the historyList is only needed when we are going to look at the year that satisfies the request, so no need to get a list of files for each year that we are simply going to remove the directory. 

See https://github.com/cmwshang/pgbackrest/commit/a5cd80d36d5adfbaa242031cd1defa8e32564228 for changes and updates to tests to incorporate if you agree with the modifications. 

@dwsteele and @pgstef - I kept the currentBackupList and historyList Descending since it is a pattern we use when expiring files to expire the dependents first in the event something happens during file removal, we still have the Full - although I don't see how that would be helpful since the history files are useless if the backup is removed and by removing a dependent first, we would leave the history incorrect if an error occurs during the removal of a dependent.

Reply:
> t would be more efficient if the year list was sorted Ascending. For example, if a user would like to keep the last 2 years of history then we can stop looking when we get to the current-year - 2 years.

That looks good to me.

> Also, the historyList is only needed when we are going to look at the year that satisfies the request, so no need to get a list of files for each year that we are simply going to remove the directory.

Very good point -- saving IO is always a worthwhile.

> although I don't see how that would be helpful since the history files are useless if the backup is removed

For the purposes of history, each manifest is valuable by itself. For instance, if you wanted to do trend analysis you don't need every manifest, maybe just one per month or one per week max. For these purposes full, diff, incr does matter because you can still measure total database size and WAL generated per day, probably the most important metrics. Anyway, if there are gaps because of failed deletes they would be outside the retention period anyway so it doesn't seem very important.

Reply:
Changed does to doesn't above.

Reply:
The suggested change looks good to me too. Thanks @cmwshang  :-)